crates/lang/ir/src/ir/chain_extension.rs

```diff
@@ -0,0 +1,734 @@
+    item: syn::TraitItemMethod,
+    id: ExtensionId,
+}
+
+impl ChainExtensionMethod {
+    /// Returns the Rust attributes of the ink! chain extension method.
+    pub fn attrs(&self) -> Vec<syn::Attribute> {
+        let (_, attrs) = ir::partition_attributes(self.item.attrs.iter().cloned())
+            .expect(
+            "encountered unexpected invalid attributes for ink! chain extension method",
+        );
+        attrs
+    }
+
+    /// Returns the span of the ink! chain extension method.
+    pub fn span(&self) -> proc_macro2::Span {
+        self.item.span()
+    }
+
+    /// Returns the identifier of the ink! chain extension method.
+    pub fn ident(&self) -> &proc_macro2::Ident {
+        &self.item.sig.ident
+    }
+
+    /// Returns the method signature of the ink! chain extension method.
+    pub fn sig(&self) -> &syn::Signature {
+        &self.item.sig
+    }
+
+    /// Returns the unique ID of the chain extension method.
+    pub fn id(&self) -> ExtensionId {
+        self.id
+    }
+}
+
+/// The unique ID of an ink! chain extension method.
+///
+/// # Note
+///
+/// The ink! attribute `#[ink(extension = N: u32)]` for chain extension methods.
+///
+/// Has a `func_id` extension ID to identify the associated chain extension method.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ExtensionId {
+    index: u32,
+}
+
+impl ExtensionId {
+    /// Creates a new chain extension method ID from the given `u32`.
+    pub fn from_u32(index: u32) -> Self {
+        Self { index }
+    }
+
+    /// Returns the underlying raw `u32` index.
+    pub fn into_u32(self) -> u32 {
+        self.index
+    }
+}
+
+impl TryFrom<syn::ItemTrait> for ChainExtension {
+    type Error = syn::Error;
+
+    fn try_from(item_trait: syn::ItemTrait) -> core::result::Result<Self, Self::Error> {
+        idents_lint::ensure_no_ink_identifiers(&item_trait)?;
+        Self::analyse_properties(&item_trait)?;
+        let methods = Self::analyse_items(&item_trait)?;
+        Ok(Self {
+            item: item_trait,
+            methods,
+        })
+    }
+}
+
+impl ChainExtension {
+    /// Returns `Ok` if the trait matches all requirements for an ink! chain extension.
+    pub fn new(attr: TokenStream2, input: TokenStream2) -> Result<Self> {
+        if !attr.is_empty() {
+            return Err(format_err_spanned!(
+                attr,
+                "unexpected attribute input for ink! chain extension"
+            ))
+        }
+        let item_trait = syn::parse2::<syn::ItemTrait>(input)?;
+        ChainExtension::try_from(item_trait)
+    }
+
+    /// Analyses the properties of the ink! chain extension.
+    ///
+    /// # Errors
+    ///
+    /// - If the input trait has been defined as `unsafe`.
+    /// - If the input trait is an automatically implemented trait (`auto trait`).
+    /// - If the input trait is generic over some set of types.
+    /// - If the input trait's visibility is not public (`pub`).
```

Comment:
`"ink! chain extensions with supertraits are not supported, yet"` is a missing error which can happen.